protocol.go

```diff
@@ -106,10 +108,12 @@ func NewConn(conn net.Conn, opts ...func(*Conn)) *Conn {
 func (p *Conn) Read(b []byte) (int, error) {
 	p.once.Do(func() {
 		p.readErr = p.readHeader()
+		p.conn.SetReadDeadline(time.Time{})
```

Comment:
Move to `readHeader`

Reply:
I had tried and am trying again and the new tests introduced in this PR just block.

Reply:
Can you run your branch on your side with the tests that I introduce here, please?

Reply:
Actually, in your code you're setting a deadline in the connection in a place where the connection has been read already and the `p.bufReader` is available so that's why it wouldn't work.

Reply:
I wondered if this ends up somewhat reintroducing the problem that #74 tried to mitigate. At this point, a connection can just sit forever and eat up resources after an initial read. But I think it's fine. At this point the calling client has decided that it is in fact a valid connection (based on the parameters passed in to this lib with acceptance criteria) and so by the time this read is called it's really on the calling client to decide when the connection should be terminated.

That said there is something maybe subtle here: if the client has called `SetReadDeadline` earlier, then the first call to `Read` will now overwrite their set deadline.

If you agree this can be an issue, it may be worth using an atomic value that is empty unless SetReadDeadline has been explicitly called, at which point the atomic value is set with the parameter value. When this once function is run, only set the read deadline to infinity if that atomic value is empty (not zero, as they could have passed in zero; or, if you want to avoid pointers, in SetReadDeadline, you could record them passing in a zero as a negative value, and check that here). Otherwise set it to the value in the atomic value.

As for a default -- my suggestion would be one or two seconds. For most software in a datacenter or cloud that's an eternity, and if you're using this with e.g. mobile apps you can set it higher to account for disadvantaged networks.

Reply:
@pires From my point of view, that's because what you're testing is the libs ability to close a connection for a user, when that's not what you should be testing. What I thought should be tested is the libs ability to reject connections if the policy requires proxyproto, otherwise, it should silently fail and allow the user to handle the timeout. It should just not block the read for the user if a the header is not set. Here's a PR (#80) including the changes I mentioned as well as the tests I thought we should have. Am I interpreting this wrong?

Reply:
cc @igor-kupczynski @bohanyang who also have been contributing recently and may have some time to share their thoughts. More opinions will help my brain reach a consensus 😅 thank you all, once again.

Reply:
Needed to think for a bit. I don't have a good answer for this PR yet, but FWIIW we internally wrap the `go-proxyproto` with a `sync.Once` and then use `proxyproto.ReadTimeout` within that.


Simplified:
```go
readHeadersOnce.Do(func() {
  ..
  hdr, err := proxyproto.ReadTimeout(buf, timeout)
  ..
  if err != nil {
    if err == proxyproto.ErrNoProxyProtocol {
      // we explicitly handle `NoProxyProtocol`
      ..
    }
    ..
  }
  // do something with hdr
  ..
}
```

We also use `SetDeadline`, `SetReadDeadline` in some parts of the codebase.

We use 30 secs for the timeout, which is probably an overkill and now that I think about that maybe we should lower it.

Anyways, will take another look in the evening / tomorrow and see if I have any concrete suggestions here.

Reply:
OK, I think I've finally loaded some more context into my head ;) so we don't use the `proxyproto.Listener` wrapper. The origins of this decision are clouded in history, but I suspect they have something to do with the fact that we already customize the listeners and connection. That said, #74 and #76 are unlikely to affect us in any way :)

Back to the topic at hand. The change in #74 may be problematic in case someone whats to setup their own deadlines on the connection (because we overwrite them https://github.com/pires/go-proxyproto/pull/74/files?file-filters%5B%5D=.go#diff-7a50e7dfed70deb86642d34ce85399fa54d3402304c3fd31157a4552a163ad38R56-R58). Not sure how often ppl may want to do this.

As per https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/
> Deadlines are not timeouts. Once set they stay in force forever (or until the next call to `SetDeadline`), no matter if and how the connection is used in the meantime. **So to build a timeout with `SetDeadline` you'll have to call it before every Read/Write operation**.

I think it would make sense then to follow @antoniomika suggestion and move this `p.conn.SetReadDeadline(time.Time{})` to `readHeader`.  (It fails `TestCopyFromWrappedConnection` with a nil pointer, probably because of https://github.com/pires/go-proxyproto/blob/37fda369c74f0c0dac2c99f52f9d5761bcb17822/protocol_test.go#L876 but I'm guessing this can be fixed. Otherwise the tests doesn't block.


I wonder, would it make sense to store the timeout in an extra field in Listener (we already do because of #74) and then if set use `ReadTimeout` instead of `Read` ? I probably miss something, as the test now hangs, but what I have in mind is shown below:
```diff
Index: protocol.go
===================================================================
diff --git a/protocol.go b/protocol.go
--- a/protocol.go	(revision bab7dd263f970c6c7b509e39ecfa2d50560f3f31)
+++ b/protocol.go	(date 1630424364577)
@@ -30,6 +30,7 @@
 	ProxyHeaderPolicy Policy
 	Validate          Validator
 	readErr           error
+	readHeaderTimeout time.Duration
 }
 
 // Validator receives a header and decides whether it is a valid one
@@ -45,6 +46,15 @@
 	}
 }
 
+// WithTimeout adds ...
+func WithTimeout(t time.Duration) func(*Conn) {
+	return func(c *Conn) {
+		if t != 0 {
+			c.readHeaderTimeout = t
+		}
+	}
+}
+
 // Accept waits for and returns the next connection to the listener.
 func (p *Listener) Accept() (net.Conn, error) {
 	// Get the underlying connection
@@ -53,12 +63,6 @@
 		return nil, err
 	}
 
-	if d := p.ReadHeaderTimeout; d != 0 {
-		// The deadline will be reset after parsing the header.
-		// Otherwise, future p.conn.Read() will timeout.
-		conn.SetReadDeadline(time.Now().Add(d))
-	}
-
 	proxyHeaderPolicy := USE
 	if p.Policy != nil {
 		proxyHeaderPolicy, err = p.Policy(conn.RemoteAddr())
@@ -73,6 +77,7 @@
 		conn,
 		WithPolicy(proxyHeaderPolicy),
 		ValidateHeader(p.ValidateHeader),
+		WithTimeout(p.ReadHeaderTimeout),
 	)
 	return newConn, nil
 }
@@ -108,7 +113,6 @@
 func (p *Conn) Read(b []byte) (int, error) {
 	p.once.Do(func() {
 		p.readErr = p.readHeader()
-		p.conn.SetReadDeadline(time.Time{})
 	})
 	if p.readErr != nil {
 		return 0, p.readErr
@@ -215,7 +219,16 @@
 }
 
 func (p *Conn) readHeader() error {
-	header, err := Read(p.bufReader)
+	var (
+		header *Header
+		err    error
+	)
+	if p.readHeaderTimeout != 0 {
+		header, err = ReadTimeout(p.bufReader, p.readHeaderTimeout)
+	} else {
+		header, err = Read(p.bufReader)
+	}
+
 	// For the purpose of this wrapper shamefully stolen from armon/go-proxyproto
 	// let's act as if there was no error when PROXY protocol is not present.
 	if err == ErrNoProxyProtocol {
```

Reply:
Thanks @igor-kupczynski! Take a look here, I think I covered those worries: https://github.com/pires/go-proxyproto/pull/80

Reply:
Nice, what do you think of using the `proxyproto.ReadTimeout` instead of the stdlib `Deadline`s?

(BTW I'm not gating anything here, mostly just thinking aloud; the final call goes to Pires).

Reply:
I think personally, I'd prefer using the deadlines in this scenario. Main reason being, I like the idea of this lib (using a listener) acting as a transparent layer which populates the right addresses (ergo it is a drop in replacement) rather than requiring more implementation work. Do you foresee an issue using deadlines if we retain the passed in values?

Reply:
> Do you foresee an issue using deadlines if we retain the passed in values?

Not really, on a quick check #80 does what it says on the tin.

`ReadTimeout` is already implemented in the lib so I'm thinking it maybe simpler to reuse it inside the `readHeader` (see my diff above). Anyways, since I'm not using the lib via the listener, but rather the `ReadTimeout` directly, I think I'm not the target "client" there :)

Reply:
I don't think there's anything wrong with that. I just figured deadlines would be the preferred way to implement this with minimal goroutines. Ultimately, we're still in line with the synchronous methods recommendation (https://github.com/golang/go/wiki/CodeReviewComments#synchronous-functions) so I think either works!

Reply:
Thank you all for your invaluable input and help. #80 is highly likely to close this PR and i'll cut a release immediately after it has been merged.